### PR TITLE
Add a code of conduct doc.

### PR DESCRIPTION
**Pull request by jdalton**
Adding a code of conduct to help [avoid issues](https://twitter.com/jashkenas/status/704399369051770880) over moderation and behavior. This will spell out what is and what isn't acceptable to help foster a better environment and prevent infighting on how moderation should be handled.

### CODE REVIEW

**Comment by koglerjs** [comment]
Interesting approach.  

So we should really only be interested in this Code of Conduct insomuch as how we might be able to use it to restrain your behavior.  By your own indication.  Yes?  That's why you linked to the report of your misbehavior, correct?  

I am concerned by the report of your editing the comments of other users (I didn't even know Github provided that tool).  Play the game of saying what you mean and apologizing by editing afterwards if you like.  Enforcing edits on others and hiding the evidence is a closed, hamfisted approach guaranteed to stir up trouble.  

I will of course be keeping a copy of this comment for comparison.

**Comment by jdalton** [comment]
> So we should really only be interested in this Code of Conduct insomuch as how we might be able to use it to restrain your behavior. By your own indication. Yes

It's more to iron out what is acceptable and what is not. The link was to give context. Others may have a different bar/criteria for moderation. Having a code of conduct helps set that bar and outlines what is acceptable behavior.

> I am concerned by the report of your editing the comments of other users (I didn't even know Github provided that tool). Play the game of saying what you mean and apologizing by editing afterwards if you like.

I tend to side more with stopping the derailing of threads and keeping nonconstructive bits out. Again though a CoC will set the expectations of behavior and conduct. The one I've provided is what I've used for other projects.

**Comment by craigmichaelmartin** [comment]
It's a sad day that something like this may even be needed, IMHO.
> I tend to side more with stopping the derailing of threads and keeping nonconstructive bits out.

My email begs to differ :P

**Comment by jdalton** [comment]
> It's a sad day that something like this may even be needed, IMHO.

Code of conducts are a good thing.

> My email begs to differ :P

It's unfortunate that disagreements and spats happen but they are usually dealt with and corrected in back channel or cleaned up after cooler heads prevail. Public bickering doesn't serve the good of the project.

**Comment by jdalton** [comment]
I think we all participate in these projects to make things better.

**Comment by craigmichaelmartin** [comment]
> Code of conducts are a good thing.

I'm not sure of any great open source project that originated from a code of conduct. Code of conducts always come after a sad day, whether experienced first-hand or not. They are good insofar as there are, sadly, folks who need them.

> I think we all participate in these projects to make things better.

No doubt, thank you for your contributions!!

**Comment by jdalton** [comment]
> I'm not sure of any great open source project that originated from a code of conduct. Code of conducts always come after a sad day, whether experienced first-hand or not. 

I guess I disagree. A code of conduct is a win for moderators and the community.
I think @thealphanerd can expand a bit more on that.

**Comment by samccone** [comment]
> A code of conduct is a win for moderators and the community.

:+1: 

editing peoples comments because of bickering 

:-1: 

If a maintainer edited someones comment because it was noise / bickering I would say that would be in violation of the CoC and would be terms for removal from the project. 


+1 to merge

**Comment by jdalton** [comment]
> (just to be extra clear I appreciate everyones time and work that they put into this project, my above comment only aims to support a healthy community on both sides of the proverbial fence here)

I'm cool with it. I think clarification and outlining things is a good thing.

**Comment by kittens** [comment]
:+1: to a code of conduct, discussion on specific scenarios that violate it are probably out of scope for this issue though. Although I do find it pretty ironic that the one pushing for a CoC is the one being accused of violating it.

**Comment by jdalton** [comment]
> Although I do find it pretty ironic that the one pushing for a CoC is the one being accused of violating it.

It's my hope that it clears things up for the future.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+Contributor Covenant Code of Conduct
```
Comment (megawac) [inline CODE_OF_CONDUCT.md]: what is the source of this code of conduct?

Comment (jdalton) [inline CODE_OF_CONDUCT.md]: > what is the source of this code of conduct?

The last line covers it.

> This Code of Conduct is adapted from the Contributor Covenant, version 1.4, available at http://contributor-covenant.org/version/1/4.

Comment (megawac) [inline CODE_OF_CONDUCT.md]: Ah whoops, missed that. Mind just fixing [the formatting](https://github.com/jdalton/underscore/blob/coc/CODE_OF_CONDUCT.md)

Comment (jdalton) [inline CODE_OF_CONDUCT.md]: Done.

**Comment by TheAlphaNerd** [comment]
> It's a sad day that something like this may even be needed, IMHO.

It is really unfortunate you feel that way. I'm not 100% of the situation that has brought this up, but I can give some thoughts based on the experience with a CoC and moderation policy form the Node.js core team.

Having a CoC is a baseline, although it often gets brought up during a time of turmoil. A CoC is a signal, it is a way to let people know that your project cares about protecting individuals from harassment. There really is not a valid reason to not want to protect individuals from harassment.

That being said, a CoC is completely useless without
  *  A clear moderation policy
  *  Buy in from the collaborators

A clear moderation policy allows collaborators to moderate appropriately and not get stuck bike shedding about process. This has been experienced first hand in the node project when we had a CoC without clear moderation policies, and we spent days figuring out how to deal with a situation while members of our community were being actively harassed.

Our moderation policy [can be found here](https://github.com/nodejs/TSC/blob/master/Moderation-Policy.md)

The code of covenant comes with a paragraph outlining that moderation may happen, but it is not clear about how to moderate, or how to document the process. The node project has a different process for moderating content from a [collaborator vs. a non-collaborate](https://github.com/nodejs/TSC/blob/master/Moderation-Policy.md#collaborator-posts), it sounds like being explicit about this might be important for this project.

It is sad that CoC's and moderation policies often come up after an unfortunate instance makes it seem necessary, but that does not mean that having a CoC itself is toxic. Having a clear CoC and moderation guideline will reduce the odds of these types of events happening in the first place.

We use linters to make sure our code fits within the agreed upon standards of a project. We use a CoC to make sure our behavior does as well.

It seems to me like @megawac @akre54 @michaelficarra are the ones mostly merging code ATM... if we do not have buy in from these individuals a CoC is worse than useless.

If anyone has questions about this type of stuff they are not comfortable asking in the open please feel free to DM me. Individuals with power in the world have a duty to empower and protect those who do not.

**Comment by koglerjs** [comment]
First off, I'm in favor of CoCs generally for most if not all projects.  

And to be further clear, I don't believe I have the "full story" (nor am I interested in knowing it) and certainly entertain the notion that multiple parties have taken damaging actions.  

One can imagine your (@jdalton) using a code of conduct to _excuse_ editing others' comments.  It would be an act of deep political cynicism to use a conflict whose creation is at least partially your responsibility to pass a code of conduct that would allow you greater leeway in resolving conflict in your favor.  It would be unfortunate if people got this impression.  

Since that is not what's happening here, you're presenting the code of conduct as a means of considering action against yourself or one contributing to conflict in a similar manner as yourself in a future situation.  I am reading this correctly, right?  

In that case a good way to build the argument for a Code of Conduct would be for you to apply that code to the preceding situation or some abstract of it.  

Some person with admin powers uses those powers in a way that makes another contributor feel unwelcome and _generates more conflict_ (as witnessed by your link).  The community recourse would be to contact jashkenas and jashkenas would then be enabled by the code of conduct to boot ~~you~~ that admin from further responsibility, or otherwise limit moderation privileges.  

Ths is a slight improvement over the various privilege/permission wars I've heard of, in passing, between you and jashkenas, I agree!  

>It's my hope that it clears things up and prevents issues in the future.

I do not think you have done a very good job explaining _why_ this is your hope with the specifics of this project.

**Comment by jdalton** [comment]
> First off, I'm in favor of CoCs generally for most if not all projects.

:clap:

> Some person with admin powers uses those powers in a way that makes another contributor feel unwelcome and generates more conflict (as witnessed by your link). 

This was a squabble between a few core contributors and the tweet was one side of that squabble. I don't care to turn this into a TMZ event so rather start the process to help prevent things like this in the future. 

A code of conduct and a moderation policy is a good start.

**Comment by koglerjs** [comment]
You haven't even demonstrated the capacity to work on things like this in the present.

**Comment by jdalton** [comment]
> You haven't even demonstrated the capacity to work on things like this in the present.

Over the last four years I've contributed bug fixes, api design, triaged issues, started the merge discussions, and have worked with and incorporated several core into semi daily discussions on libs, utils, projects to make things better for developers. And now I'm working on this.

**Comment by jdalton** [comment]
It's not easy for sure and sometimes there's bumps but over the years things trend in a positive direction.

**Comment by koglerjs** [comment]
Starting a pull request for a CoC does not disavow you of responsibility for the circumstances that you **directly linked to** as justification for the CoC.  If you did not wish to bring them up and consider them **you should not have brought them up**.  

Your overall actions as a contributor to open source is (was) not up for consideration.

**Comment by jdalton** [comment]
> Starting a pull request for a CoC does not disavow you of responsibility for the circumstances that you directly linked to as justification for the CoC

I'm fine with folks having a different opinion on how to best moderate. I think we can both agree that outlining a CoC and moderation policy is a good thing.

**Comment by jdalton** [comment]
> Your overall actions as a contributor to open source is (was) not up for consideration.

Oh that was contributions to Underscore. But ok.

**Comment by koglerjs** [comment]
My mistake.  I am more or less a bystander.  

As a bystander, my impression of your involvement with the underscore project is overwhelmingly that you are someone that stirs up trouble.  Certainly you and jashkenas have had your differences.  Certainly you are not entirely to blame for dramatic happens.  

It does not help your case, in my opinion, to confine your demonstration of authority to the Underscore repository.  I preferred thinking of you as resting on your authority as the maintainer and core collaborator of Lo-dash.  (Not that I think that helped, either, of course.)  

As far I'm concerned, this very topic is an instance of your inability to de-escalate conflict.  In fact I think you have no small ability to propagate it, which would be much better handled if you could acknowledge that, and act as a humble human being under the judgment of any code of conduct you might have proposed.  

I began this conversation with bewilderment.  I don't understand you.  Or I understand you perfectly.  There's not a lot of ambiguity in this conversation on my end.  

You've settled on the CoC as a perfect middle ground that no one can contest, and having claimed it, you can't be attacked.  You should perhaps consider that you do damage to the credibility of CoCs when you initiate them in such a strange way.

**Comment by jridgewell** [comment]
Guys, can we not start an argument in the PR _for_ a CoC? That's ridiculous.

This a positive change, I'm in favor. @TheAlphaNerd's comment seems particularly relevant:

> The node project has a different process for moderating content from a [collaborator vs. a non-collaborate](https://github.com/nodejs/TSC/blob/master/Moderation-Policy.md#collaborator-posts), it sounds like being explicit about this might be important for this project.

That'd make this even better.

**Comment by koglerjs** [comment]
Obviously it is not my place, as a non-contributor, and only a long-time user of Underscore and Lodash, to stop a CoC from being instated.  I would not want any of my comments to be read as such.

**Comment by megawac** [comment]
Anyway, I'm :+1: for the COC in general.

> The code of covenant comes with a paragraph outlining that moderation may happen, but it is not clear about how to moderate, or how to document the process. The node project has a different process for moderating content from a collaborator vs. a non-collaborate, it sounds like being explicit about this might be important for this project.

I agree due to the circumstances prompting this

**Comment by jdalton** [comment]
@koglerjs 
> As a bystander, my impression of your involvement with the underscore project is overwhelmingly that you are someone that stirs up trouble. Certainly you and jashkenas have had your differences. Certainly you are not entirely to blame for dramatic happens.

The daily ins and outs are much more mundane. 

> It does not help your case, in my opinion, to confine your demonstration of authority to the Underscore repository. I preferred thinking of you as resting on your authority as the maintainer and core collaborator of Lo-dash. (Not that I think that helped, either, of course.)

It's certainly a tricky position to be in. Being 100% isolated is hard and doesn't help folks wanting to make plugins or projects that can leverage either project. It takes a lot of cross-lib collaboration on APIs, behaviors, compat paths. And sometimes lines get blurred and toes get stepped on but as I've written earlier I think we all participate in these projects to make things better.

@jridgewell @megawac 
> This a positive change, I'm in favor. @TheAlphaNerd's comment seems particularly relevant:

:+1:

**Comment by jdalton** [comment]
@jridgewell @megawac

I'm :+1: on the moderator policy. Would that be more approp for a separate PR for core to hash out?

**Comment by koglerjs** [comment]
>The daily ins and outs are much more mundane. 

I'm sure they are.  Something I had typed out but did not commit: months of steady sailing do not make up for the destruction of a mast.

**Comment by koglerjs** [comment]
(Which wouldn't matter if you were not here talking about building masts.)

**Comment by jdalton** [comment]
It is what it is. I'm not throwing in the towel just yet.

**Comment by jdalton** [comment]
We'll work through it and end up in a better place.

**Comment by koglerjs** [comment]
These are vagaries.  Say something real?

**Comment by jdalton** [comment]
> These are vagaries. Say something real?

There are plenty of places to help out if you want to make things better too. From docs, to improving test coverage, to ironing out a release changelog, to identifying first-time PR issues, etc.

**Comment by koglerjs** [comment]
Yes, you are within your right to defer me to participation invitation.  

I like to help out by providing incidental political clarification.  (Or lack thereof.)

**Comment by akre54** [comment]
@megawac I'm not sure the wisdom of employing the very comment moderation policy we're debating in this thread. None of the removed posts appear to be especially egregious.

Deleting or editing "off topic" or "unconstructive" comments is a really slippery slope. Not least because it's incredibly subjective.

**Comment by megawac** [comment]
I suppose, however this thread was getting lengthy at 25 posts in one day and none of the 8 posts I deleted (after providing a warning that posts off topic from the COC topic would be deleted) discussed the topic of the thread. I wanted to avoid further derailment.

Anyway now -- **yay or nay for a COC? Any changes to the proposed COC?**

**Comment by akre54** [comment]
I think the points raised are important to discuss because they provide context. Particularly for something as nuanced as a Code of Conduct, it's important to get the details right.

**Comment by akre54** [comment]
A code of conduct isn't a magic bullet. We're talking repeated and flagrant violations of respectful tone and unwillingness to handle disagreements.

**Comment by jdalton** [comment]
> A Code of Conduct isn't a magic bullet.

That plus a moderation policy seems like a start. 

> We're talking about repeated and flagrant violations of respectful tone and unwillingness to handle disagreements

I think most core will agree you've done your part to stir up trouble. Having a code of conduct will help outline what's accepted and what's not and a moderation policy will help with the moderation of core and users.

**Comment by koglerjs** [comment]
>I suppose, however this thread was getting lengthy at 25 posts in one day and none of the 8 posts I deleted (after providing a warning that posts off topic from the COC topic would be deleted)

Uh, you **edited your warning in** and I didn't notice until after the comments were suddenly gone.  

Github's tools for this are terrible.  

I don't mind particularly that my comments were removed.

**Comment by koglerjs** [comment]
Though I do think it was hamfisted, overbearing moderation.  As akre54 noted, it's a little subjective; why is it that jdalton was given the last word even though his comment was after the cutoff?  These are the questions and criticisms you open yourself up to; conflict could escalate if this were done to someone who cared a little more.  

The point I'm trying to make is that a really good moderator policy won't help if the tools don't support it.  And there we might be stuck.

**Comment by jridgewell** [comment]
> @jdalton: It is what it is. I'm not throwing in the towel just yet.


> @koglerjs: These are vagaries.  Say something real

Those are both _full_ comments. Who really cares who gets that last word in that conversation? That's not acceptable.

**Comment by koglerjs** [comment]
I wouldn't repost deleted comments.  jdalton and I have let them lie.  

My point isn't that there's any disastrous miscarriage of justice here.  It's that justice requires a lot of thought.  

Also I didn't want to sound down on a moderation policy; I'm for one, just a little cynical of Github.

**Comment by michaelficarra** [comment]
Let's please please please keep this discussion about the content of the PR and not previous violations of a to-be-defined code of conduct. I believe @jdalton when he says he has good intentions with the introduction of a CoC, and I think he now realises how the community has been affected by his actions.

I agree with @TheAlphaNerd that we need to be careful to outline a moderation policy, though maybe one that is not as formal as node's. In fact, I think that will be the most important outcome of this effort. I have personally stuck to the following guidelines when moderating in the past:

* edits to your own comment (besides typographical/formatting fixes) after someone has responded should be append-only and clearly marked using **edit:**
* deletions of your own comment are only allowed before someone has responded
* edits of someone else's comment are only allowed
  * to edit out violations of the CoC (clearly marked with **edit:**, moderator's username, timestamp, and rationale)
  * when it is used to direct future readers to a relevant point later in the conversation, usually the resolution (append-only, clearly marked with **edit:**, moderator's username, and timestamp)
* deletions of someone else's comments are never okay

**Comment by michaelficarra** [comment]
@megawac Yes. See the point above that about editing someone else's comments.

**Comment by koglerjs** [comment]
>I may of corrected the formatting or something within 5 minutes of posting the warning but thats it.

In an active thread that's enough.  Not a big deal, seriously!  Just a point of order that I was not intentionally ignoring your warning or something.  

I too agree strongly with @michaelficarra's guidelines.  It might be worth saying that _any_ edit, no matter how close to the original post time, should be noted, with some discretion w.r.t typos.  And a moderator proclamation might deserve its own comment.

**Comment by jridgewell** [comment]
@michaelficarra: That sounds fine to me.

**Comment by akre54** [comment]
Seems reasonable. Are all Code of Conduct violations subject to editing or only egregious abuses (i.e. hate speech, threats of violence, etc)?

**Comment by michaelficarra** [comment]
I would like to add that deletions of comments whose entire content is either "+1", ":+1:", or "me too" is perfectly fine by me :smiley:.

@akre54 I don't see why any CoC violation is more serious than another. If that is the case, maybe a CoC shouldn't restrict the "less serious" content. The only difference I can see in the severity of response would be a one-time offense versus repeat offenses.

**Comment by TheAlphaNerd** [comment]
Self editing a comment after it has been read is important. In node we often give a person a heads up that they  should consider self moderating, and that often involves going back and deleting / modifying comments to be on track and non derailing.

If an individual refuses to self moderate, or the comments are really offensive (which has happened more than once) than we delete the comments and keep a screen cap of the text. We maintain a private moderation repo where we keep track of all moderation instances.

If certain threads are really active / toxic (as was the case with the [chakra PR](https://github.com/nodejs/node/pull/4765)) we assigned an individual to monitor the thread who moderated / deleted comments without much hesitation.

People do not have an inherit right to take part in an open source project. it is a privilege. People also don't have an inherit right to start airing personal derailing opinions or philosophy in a thread and expect to not be moderated. The way in which this was handled in this thread for example was very much inline with this, a stern warning was given with an opportunity to self moderate, followed by moderation when the individual did not comply. 

If a comment is deleted and an individual feels their voice is being "silenced" they have every opportunity to find a new way of expressing their ideas in a way that does not violate the CoC or general decency. if they find themselves unable to find a way to do so, they will find themselves moderated and potentially banned.

**Comment by akre54** [comment]
> I would like to add that deletions of comments whose entire content is either "+1", ":+1:", or "me too" is perfectly fine by me :smiley:.

Sounds good to me :+1:

> In node we often give a person a heads up that they should consider self moderating, and that often involves going back and deleting / modifying comments to be on track and non derailing.

This is a great idea. Does node have a policy for repeat offenders?

> they have every opportunity to find a new way of expressing their ideas in a way that does not violate the CoC or general decency

Not that I'm disagreeing but "general decency" is a *really*, *really* subjective thing. I think we should try to be courteous towards each other and always try to assume good faith, but leaving moderation by intent / "general decency" up to a fallible moderator is asking for trouble.

**Comment by koglerjs** [comment]
>The way in which this was handled in this thread for example was very much inline with this, a stern warning was given with an opportunity to self moderate, followed by moderation when the individual did not comply. 

This is not at all what happened.  I was not the only one off topic, I wasn't warned effectively, and there was no 'encouragement' to 'self moderate.'  Trying to hold this thread up as an example of what good moderation might look like weakens the case for these otherwise sensible rules.  

Bootstrapping the moderation policy like this might be a good idea, but let's not pretend it was somehow retroactively effective, especially by stretching the interpretation of events.

**Comment by TheAlphaNerd** [comment]
> but "general decency" is a really, really subjective thing
¯\_(ツ)_/¯

Totally is. That being said we've found it is better to trust the judgement of those who we collaborate with. Things are documented, and if we believe someone was being hasty we definitely call them out on it. Then we improve moving forward.

It is also worth mentioning that we do not often opt to delete, especially when things have a troll like tone. We have had more than one instance in the past where we have take the time to talk the individual down, and eventually they self moderated (many hours later). This is to avoid feeding trolls, making people feel marginalized, or potentially increasing the potential for abuse.

@koglerjs I've been trying really hard to not engage with you, but I'm going to say this much. You do not come across as someone acting in good faith. You are consistently derailing this thread, playing devil's advocate, and generally not offering constructive feedback. I want to believe that you have good intentions, but you should be aware that you come off very toxic to me, and likely others. I strongly encourage you to take a look at what you are doing, and consider how you could express yourself in a constructive way that will work towards a meaningful solution. Contributing and taking part in open source is a privilege, not a right, and repeated derailing behavior is an easy way to marginalize your voice and ostracize yourself from communities

**Comment by akre54** [comment]
> You do not come across as someone acting in good faith. You are consistently derailing this thread, playing devil's advocate, and generally not offering constructive feedback.

I strongly disagree. With the exception of a few minor squabbles in the middle of the thread (since deleted), I think he's been totally fair given the circumstances. Again, subjective. What in what he said do you see as derailing?

**Comment by TheAlphaNerd** [comment]
@akre54 I'm going to avoid derailing this thread myself... I likely should not have even engaged the way I did. I apologize for not holding myself to a higher standard.

**Comment by akre54** [comment]
Thank you for self-moderating, and setting a good example. I am curious to hear your thoughts off-thread, either in Gitter or via email.

**Comment by koglerjs** [comment]
>I likely should not have even engaged the way I did

You are **too right**.  Thank you, sincerely, for publicly indicating as such, it allows me to shorten this comment and keep its combativeness low.  

I will note that you do not address my criticism of your use of this thread's moderation as an example, and have instead chosen to make an assessment of me generally, which I will of course consider.  

I promise this is on topic: the course of events is that JDD suggests bringing in a Code of Conduct, referencing incidences of his reported/alleged misbehavior.  He summons @TheAlphaNerd, who, acting initially as a reference for similar communal management, eventually puts on his moderator attitude rather strongly towards me.  I don't take it personally, I assure you.  

But how amusing is it that _I_ end up on trial, at the close of this topic!  I would say that I had played the politics poorly, except I am not really a community member (just an interested party) and cannot really fear my voice being 'marginalized.'

**If this is meant to be a demonstration of how the community should function under a Code of Conduct with moderation rules, how successful has it been?**  You have conveniently ignored the actions of your friend jdalton, @TheAlphaNerd, and the way you swiveled from considering my argument to evaluating my toxicity as a person makes me think that maybe every user starts looking like a troublemaker to a moderator, except of course for those that are politically aligned with the moderator.  

You are even more of an outsider to this community than I am, so far as I can tell.  

My toxicity is direct and impersonal.  I know I was pushing @jdalton.  I wanted to see what he had to say for himself.  If he had been able to say something like "This CoC could have been used to limit some of my recent behavior"--_anything_ approaching willingness to express some public awareness of his own responsibility!--I would have been much less aggressive in my inquiry and appreciative that the proposal was in good faith.  My line of reasoning had a purpose, and I'm not really sorry about it (for that reason I'm also not at all bothered by the deletion).  

I'm still in favor of the CoC on principle and the moderation rules because they would have helped this topic.  But I don't think they make moderation easy or successful.  

This will be my last post; I'll read and consider responses but feel like I've said what I needed to say.  Apologies for any remaining stepped-on toes I may have clomped on.

**Comment by akre54** [comment]
Alright @koglerjs that's far enough. He apologized. Please trim your comment down.

**Comment by Rayraz** [comment]
_First off, I'm writing this in an attempt to help move things forward. If this comment does not succeed at this, feel free to delete it._

It seems to me the main reason for Underscore's presence here on GitHub is to evolve and continuously improve the quality of the underscore project. While this thread does not deal with code, it does deal with the quality of this project. As such, it saddens me to see that - imho - most of the conversation has been people trying not to devolve into bickering, but failing.

When a pull request with some code changes is not quite ready to be merged, you don't point at the developer, you point at parts of the code that could use improvement and make your case. Improving community guidelines has equally much to do with evolving and improving the quality of the project as writing cleaner code or better tests.

Let's treat the task at hand the same way this community would treat a code commit and focus on converging towards something that makes underscore better.

I would like to suggest that one of the powers that be draft a few moderation rules and / or propose changes to the CoC if needed? I think a serious and constructive attempt to materialize some concrete rules may help regain some focus here.

**Comment by paulfalgout** [comment]
I think `CONTRIBUTING.md` should link to the code of conduct.

**Comment by TheAlphaNerd** [comment]
Building on that it might be a good idea to reference it is the Readme

**Comment by jdalton** [comment]
@megawac @jridgewell Is this PR good to merge with the idea of follow ups for moderation policy and any cross referencing?

**Comment by jridgewell** [comment]
Good with me, but I'll wait for another :+1:.

**Comment by akre54** [comment]
So concretely, what are we agreeing to here by merging this?

**Comment by jdalton** [comment]
> So concretely, what are we agreeing to here by merging this?

For this PR that the code of conduct outlined is acceptable.

**Comment by jdalton** [comment]
At the time of creating the PR I wasn't aware that moderation policies were a thing so that's something for a follow up PR (but really should be hashed out by core contribs).

**Comment by akre54** [comment]
Right but concretely what does that mean. Most of the language in this change is about attempting to create a safe space and limiting harassment in the project. If we don't have clear definitions of trolling or harassment, what have we gained?

> At the time of creating the PR I wasn't aware that moderation policies were a thing

No codified policy, but a generally accepted standard of behavior the core contributors have tried to moderate by. In the past bad behavior has been called out and dealt with on a case-by-case basis. (I believe comment editing / deletion was [brought up in 2013](https://github.com/jashkenas/underscore/issues/943#issuecomment-12674098)). Is the idea to formalize these policies into a document, rather than just calling out bad faith and underhanded comments in the thread itself?

**Comment by jdalton** [comment]
> Right but concretely what does that mean. 

I believe @TheAlphaNerd [covered that above](https://github.com/jashkenas/underscore/pull/2454#issuecomment-190509120).

> Is the idea to formalize these policies into a document, rather than just calling out bad faith and underhanded comments in the thread itself?

Yes. When I opened the PR I thought the code of conduct could perform double duty. As the thread progressed a moderation policy was mentioned as something more appropriate for outlining core and contributor moderation (Node's was used as a reference). As Underscore core continues to grow (2 added in the last week) and with the lead away at the moment, formalizing these things can help prevent issues before they start.

**Comment by jdalton** [comment]
That's not to say I don't think a code of conduct is needed. I think the PR is merge ready.

**Comment by megawac** [comment]
I'll leave it to Adam  to handle this PR. He seems to be the one still with reservations. I think we should also include Michael's suggested changes in the COC. I'm indifferent if that is done now or in the future

**Comment by jdalton** [comment]
> I think we should also include Michael's suggested changes in the COC

I think a follow up commit to tweak to core taste would be appropriate or else it would just be me guessing at the exact wording or phrasing you all would prefer.

**Comment by akre54** [comment]
> The code of covenant comes with a paragraph outlining that moderation may happen, but it is not clear about how to moderate, or how to document the process. 

Couldn't we just formalize the moderation policy here in this thread like @michaelficarra has done above? I think being explicit that substantively changing the content of your post or others is out of line? Especially when it reframes the debate to look like one that didn't actually take place.

I'm in weak support of this pull request, but I almost feel like this isn't much better than having nothing at all. I'd like to wait for one of the Underscore OGs to weigh in (@jashkenas, @tgriesser, @braddunbar) but I'm fine merging if everyone else is.

**Comment by michaelficarra** [comment]
I'm fine with merging this as long as we immediately follow it up with a PR that adds my suggested moderation policy (as a separate document I guess?).

**Comment by jdalton** [comment]
> I'm fine with merging this as long as we immediately follow it up with a PR that adds my suggested moderation policy (as a separate document I guess?).

:+1:

**Comment by braddunbar** [comment]
I'm a :+1: on the pull because I think outlining these things explicitly is a good idea. That said, I'm uncertain that it will be of assistance in the specific scenarios mentioned here.  ¯\\\_(ツ)\_/¯

**Comment by TheAlphaNerd** [comment]
> I'm fine with merging this as long as we immediately follow it up with a PR that adds my suggested moderation policy (as a separate document I guess?).

Great idea. Having the moderation policy in a PR is something that can be referenced while it is being fleshed out. Better to not push things through that are not quite ready

**Comment by koglerjs** [comment]
I was planning on drafting a PR with a moderation policy as a show of good faith and alignment with the CoC, even before this topic reopened.  (Probable plan: copy-paste @michaelficarra's proposal and make some minor suggestions on some modifications.)  I won't be able to get to it until late tonight and if anyone else wanted to take initiative that's fine by me.

**Comment by jdalton** [comment]
:shipit:?

**Comment by koglerjs** [comment]
Apologies to those waiting on a moderation policy; I've begun the process but it is more than I can finish tonight.

**Comment by michaelficarra** [comment]
@koglerjs Don't worry about it. I've created https://github.com/jashkenas/underscore/pull/2463.

**Pull request merged**